src/ralph/lib/polymorphic/models.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+
+from itertools import groupby
+
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.fields.related import OneToOneField
+
+from ralph.admin.sites import ralph_site
+
+
+class PolymorphicQuerySet(models.QuerySet):
+
+    def iterator(self):
+        """
+        Override iterator:
+            - Iterate for all objects and collected ID
+            - For each ContentType generates additional queryset
+            - Returns iterator with different models
+        """
+        result = []
+        ct_map = {}
+        for obj in super().iterator():
+            ct_map[obj.content_type_id] = obj.content_type.model_class()
+            result.append((
+                obj.content_type_id, obj.pk)
+            )
+        result = sorted(result, key=lambda x: x[0])
+        result = groupby(result, lambda x: x[0])
+
+        result_query = []
+        for k, v in result:
+            model = ct_map[k]
+            polymorphic_model = getattr(model, 'polymorphic_model')
+            if polymorphic_model and model.__name__ not in polymorphic_model:
+                result_query.extend(
+                    model.objects.filter(pk__in=[i[1] for i in v])
+                )
+
+        for obj in result_query:
+            yield obj
+
+
+class Polymorphic(models.Model):
+
+    """
+    Ralph Polymorphic model.
+
+    Added content type field to model
+
+    Example:
+        >>> class Test(Polymorphic, models.Model):
+                polymorphic_objects = PolymorphicQuerySet.as_manager()
+                objects = models.Manager()
+                polymorphic_model = 'Test'
+    """
+
+    content_type = models.ForeignKey(ContentType, blank=True, null=True)
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        """
+        Save object content type
+        """
+        if not self.content_type:
+            self.content_type = ContentType.objects.get_for_model(self)
+        super().save(*args, **kwargs)
+
+    @classmethod
+    def get_related_models(cls):
+        """
+        Return related models.
+        """
+        def get_models(model_class):
+            rel_objs = model_class._meta.get_all_related_objects()
+            model_list = [
+                getattr(model_class, i.get_accessor_name())
+                for i in rel_objs if i.model != type(model_class)
+            ]
+            return [
+                model.related.related_model for model in model_list
+                if isinstance(model.related.field, OneToOneField)
+            ]
+
+        result = []
+        for related in get_models(cls):
+            result.extend(get_models(related))
+            result.append(related)
+        return result
+
+    @staticmethod
+    def get_base_ids(model, value, limit=10):
```

Comment:
this is used only in autocomplete and it should be moved there (in lib there should be only generic code, easily reusable in other applications).